deadlock_demolition.md

```diff
@@ -0,0 +1,55 @@
+---
+layout: doc
+title: "Deadlock Demolition"
+learning_objectives:
+  - Synchronization Primitives
+  - Deadlock Detection using a Resource Allocation Graph
+---
+
+## Overview
+
+This week, you will be building a library for deadlock-resistant mutex (drm) locks. A drm should behave like a pthread_mutex_t but not allow itself to be locked by a thread if the attempt to lock it would result in deadlock. We will use binary semaphore notation for your lock's functions:i.e. post is equivalent to unlocking and wait is equivalent to locking.
+
+You will be writing four functions:
+
+* `drm_t *drm_init();`
+* `int drm_post(drm_t *drm, pthread_t *thread_id);`
+* `int drm_wait(drm_t *drm, pthread_t *thread_id);`
+* `void drm_destroy(drm_t *drm);`
+
+To detect deadlock, you will need to maintain a Resource Allocation Graph and be able perform cycle detection on it. See [this page](http://cs241.cs.illinois.edu/coursebook/Deadlock#resource-allocation-graphs) for more information about Resource Allocation Graphs.
+
+**NOTE: the provided graph data structure is not thread-safe.**
```

Comment:
Could we put a high level idea of how to do this lab here, like

* Make a graph (or use graph provided)
* Init adds to graph
* Destroy removes from graph
* post adds an edge and checks for a cycle. Returns success if no cycle, otherwise removes that edge.
* wait removes an edge

Reply:
Additional comment I made on slack:
> There's a fair bit of thinking needed to figure out how to acquire the lock without essentially using the lock around the RAG as an arbitrator. 
I think we should hint about that here.

Reply:
For init and destroy I'll put those comments in the header.
Would the other points spell it out too much though?

Reply:
I did put the other points in the header as well. Tried to keep things consolidated there. 
Do the notes I added look good?